benchmark/benchmark.sh

```diff
@@ -0,0 +1,162 @@
+# Benchmarks can be finally compared using compare.jl as follows
+#     julia --project compare.jl benchmark_1.json benchmark_2.json benchmark_3.json ...
+# Note that each case benchmarks should be compared separately.
+# If a single case is benchmarked, and all the JSON files in the current directory belong to it, one can simply run:
+#     julia --project compare.jl $(find . -name "*.json" -printf "%T@ %Tc %p\n" | sort -n | awk '{print $8}')
+# which would take all the JSON files, sort them by creation time, and pass them as arguments to the compare.jl program.
+# Finally, note that the first benchmark passed as argument is taken as reference to compute speedups of other benchmarks:
+#     speedup_x = time(benchmark_1) / time(benchmark_x).
+#
+# TL;DR: Usage example
+#     sh benchmark.sh  -v "1.9.4 1.10.0-rc1" -t "1 3 6" -b "Array CuArray" -c "tgv.jl" -p "5,6,7"
+# The default launch is equivalent to:
+#     sh benchmark.sh  -v JULIA_DEFAULT -t "1 6" -b "Array CuArray" -c "tgv.jl" -p "5,6,7" -s 100 -ft Float32
+# ----
+
+
+# Grep current julia version
+julia_version () {
+    julia_v=($(julia -v))
+    echo "${julia_v[2]}"
+}
+
+# Update project environment with new Julia version
+update_environment () {
+    echo "Updating environment to Julia v$version"
+    juliaup default $version
+    # Mark WaterLily as a development package. Then update dependencies and precompile.
+    julia --project -e "using Pkg; Pkg.develop(PackageSpec(path=join(split(pwd(), '/')[1:end-1], '/'))); Pkg.update();"
+}
+
+run_benchmark () {
+    echo "Running: julia --project $args"
+    julia --project $args
+}
+
+# Print benchamrks info
+display_info () {
+    echo "--------------------------------------"
+    echo "Running benchmark tests for:
+ - Julia:        ${VERSIONS[@]}
+ - Backends:     ${BACKENDS[@]}"
+    if [[ " ${BACKENDS[*]} " =~ [[:space:]]'Array'[[:space:]] ]]; then
+        echo " - CPU threads:  ${THREADS[@]}"
+    fi
+    echo " - Cases:        ${CASES[@]}
+ - Size:         ${LOG2P[@]:0:$NCASES}
+ - Sim. steps:   ${MAXSTEPS[@]:0:$NCASES}
+ - Data type:    ${FTYPE[@]:0:$NCASES}"
+    echo "--------------------------------------"; echo
+}
+
+# Default backends
+DEFAULT_JULIA_VERSION=$(julia_version)
+VERSION=($DEFAULT_JULIA_VERSION)
+BACKENDS=('Array' 'CuArray')
+THREADS=('1' '6')
+# Default cases. Arrays below must be same length (specify each case individually)
+CASES=('tgv.jl')
+LOG2P=('5,6,7')
+MAXSTEPS=('100')
+FTYPE=('Float32')
+
+# Parse arguments
+while [ $# -gt 0 ]; do
+case "$1" in
+    --versions|-v)
+    VERSIONS=($2)
+    shift
+    ;;
+    --backends|-b)
+    BACKENDS=($2)
+    shift
+    ;;
+    --threads|-t)
+    THREADS=($2)
+    shift
+    ;;
+    --cases|-c)
+    CASES=($2)
+    shift
+    ;;
+    --log2p|-p)
+    LOG2P=($2)
+    shift
+    ;;
+    --max_steps|-s)
+    MAXSTEPS=($2)
+    shift
+    ;;
+    --float_type|-ft)
+    FTYPE=($2)
+    shift
+    ;;
+    *)
+    printf "ERROR: Invalid argument\n"
```

Comment:
1. It'd be nice to say _which_ is the invalid argument instead of letting the user guess it
2. Error messages are typically printed to stderr

```suggestion
    printf "ERROR: Invalid argument %s\n" "${1}" 1>&2
```